scipy/interpolate/_aaa.py

```diff
@@ -0,0 +1,363 @@
+        i_non_zero = wj != 0
+        zj = zj[i_non_zero]
+        wj = wj[i_non_zero]
+        fj = fj[i_non_zero]
+
+        self.support_points = zj
+        self.data_values = fj
+        self.weights = wj
+        self.errors = errvec
+
+        # Only compute poles, residues, and zeros when required
+        self._poles = None
+        self._residues = None
+        self._zeros = None
+
+    def __call__(self, z):
+        """Evaluate the rational approximation at given values.
+
+        Parameters
+        ----------
+        z : array_like
+            Input values.
+        """
+        # evaluate rational function in barycentric form.
+        z = np.asarray(z)
+        zv = np.ravel(z)
+
+        # Cauchy matrix
+        with np.errstate(invalid="ignore", divide="ignore"):
+            CC = 1 / np.subtract.outer(zv, self.support_points)
+        # Vector of values
+        with np.errstate(invalid="ignore"):
+            r = CC @ (self.weights * self.data_values) / (CC @ self.weights)
+
+        # Deal with input inf: `r(inf) = lim r(z) = sum(w*f) / sum(w)`
+        r[np.isinf(zv)] = np.sum(self.weights * self.data_values) / np.sum(self.weights)
+
+        # Deal with NaN
+        ii = np.flatnonzero(np.isnan(r))
+        for jj in ii:
+            if np.isnan(zv[jj]) or not np.any(zv[jj] == self.support_points):
+                # r(NaN) = NaN is fine.
+                # The second case may happen if `r(zv[ii]) = 0/0` at some point.
+                pass
+            else:
+                # Clean up values `NaN = inf/inf` at support points.
+                # Find the corresponding node and set entry to correct value:
+                r[jj] = self.data_values[zv[jj] == self.support_points].squeeze()
+
+        return np.reshape(r, z.shape)
+
+    @property
+    def poles(self):
+        if self._poles is None:
+            # Compute poles via generalized eigenvalue problem
+            m = self.weights.size
+            B = np.eye(m + 1)
+            B[0, 0] = 0
+
+            E = np.zeros_like(B, dtype=np.complex128)
+            E[0, 1:] = self.weights
+            E[1:, 0] = 1
+            np.fill_diagonal(E[1:, 1:], self.support_points)
+
+            pol = scipy.linalg.eigvals(E, B)
+            self._poles = pol[np.isfinite(pol)]
+        return self._poles
+
+    @property
+    def residues(self):
+        if self._residues is None:
+            # Compute residues via formula for res of quotient of analytic functions
+            with np.errstate(invalid="ignore", divide="ignore"):
+                N = (1 / (self.poles[:, np.newaxis] - self.support_points)) @ (
+                    self.data_values * self.weights
+                )
+                Ddiff = (
+                    -((1 / np.subtract.outer(self.poles, self.support_points)) ** 2)
+                    @ self.weights
+                )
+                self._residues = N / Ddiff
+        return self._residues
+
+    @property
```

Comment:
Consider using `functools.cached_property` here. It does exactly what you are doing.

Reply:
Thanks good idea